cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/task/ProvisioningTask.java

```diff
@@ -0,0 +1,140 @@
+  protected final ProgramRunId programRunId;
+  protected final int retryTimeLimitSecs;
+
+  public ProvisioningTask(ProvisioningTaskInfo initialTaskInfo, Transactional transactional,
+                          DatasetFramework datasetFramework, int retryTimeLimitSecs) {
+    this.programRunId = initialTaskInfo.getProgramRunId();
+    this.initialTaskInfo = initialTaskInfo;
+    this.transactional = transactional;
+    this.datasetFramework = datasetFramework;
+    this.retryTimeLimitSecs = retryTimeLimitSecs;
+  }
+
+  /**
+   * Execute all subtasks, persisting task info before each subtask such that this task can be re-created from the
+   * task info stored in the ProvisionerDataset.
+   *
+   * @throws InterruptedException if the task was interrupted
+   * @throws TransactionFailureException if there was a failure persisting task info
+   * @throws Exception if there was a non-retryable exception while executing a subtask
+   */
+  public void execute() throws Exception {
+    RetryStrategy retryStrategy =
+      RetryStrategies.statefulTimeLimit(retryTimeLimitSecs, TimeUnit.SECONDS, System.currentTimeMillis(),
+                                        RetryStrategies.exponentialDelay(100, 20000, TimeUnit.MILLISECONDS));
+    Map<ProvisioningOp.Status, ProvisioningSubtask> subtasks = getSubtasks();
+
+    Optional<ProvisioningTaskInfo> taskInfoOptional = Optional.of(initialTaskInfo);
+    while (taskInfoOptional.isPresent()) {
+      ProvisioningTaskInfo taskInfo = taskInfoOptional.get();
+      persistTaskInfo(taskInfo, retryStrategy);
+
+      ProvisioningOp.Status state = taskInfo.getProvisioningOp().getStatus();
+      ProvisioningSubtask subtask = subtasks.get(state);
+      if (subtask == null) {
+        // should never happen
+        throw new IllegalStateException(
+          String.format("Invalid state '%s' in provisioning task for program run '%s'. "
+                          + "This means there is a bug in provisioning state machine. "
+                          + "Please reach out to the development team.",
+                        state, programRunId));
+      }
+
+      try {
+        taskInfoOptional = Retries.callWithInterruptibleRetries(() -> subtask.execute(taskInfo), retryStrategy,
+                                                                t -> t instanceof RetryableProvisionException);
+      } catch (InterruptedException e) {
+        throw e;
+      } catch (Exception e) {
+        handleSubtaskFailure(taskInfo, e);
+        ProvisioningOp failureOp = new ProvisioningOp(taskInfo.getProvisioningOp().getType(),
+                                                      ProvisioningOp.Status.FAILED);
+        ProvisioningTaskInfo failureInfo = new ProvisioningTaskInfo(taskInfo, failureOp, taskInfo.getCluster());
+        persistTaskInfo(failureInfo, retryStrategy);
+        return;
+      }
+    }
+  }
+  // write the state of the task to the ProvisionerDataset, retrying if any exception is caught
```

Comment:
Add an empty line before the comment. Also, better to have the method comment in javadoc format.

Reply:
done